[1.9] Automated cherry pick of #58720 #57326 #60342

joelsmith opened the pull request:
Cherry pick of #58720 #57326 #60342 on release-1.9.

#58720: Ensure that the runtime mounts RO volumes read-only
#57326: Update Dashboard version to v1.8.3
#60342: Fix nested volume mounts for read-only API data volumes

Fixes https://github.com/kubernetes/kubernetes/issues/60814 for 1.9

Note for reviewers:
The Dashboard update is necessary because previous versions of the Dashboard attempted to write to the secret volume on startup, and so the read-only secret volume change prevented the Dashboard from starting. 

```release-note
Changes secret, configMap, downwardAPI and projected volumes to mount read-only, instead of allowing applications to write data and then reverting it automatically. Until version 1.11, setting the feature gate ReadOnlyAPIDataVolumes=false will preserve the old behavior. Updates dashboard version to v1.8.3 to avoid writing to read-only locations.
```

joelsmith committed:
Ensure that the runtime mounts RO volumes read-only

Add a feature gate ReadOnlyAPIDataVolumes to a provide a way to
disable the new behavior in 1.10, but for 1.11, the new
behavior will become non-optional.

Also, update E2E tests for downwardAPI and projected volumes
to mount the volumes somewhere other than /etc.

Commit:
Update Dashboard version to v1.8.3

liggitt commented on cluster/addons/dashboard/dashboard-controller.yaml:
```diff
@@ -56,8 +56,7 @@ spec:
       - name: kubernetes-dashboard-certs
-        secret:
-          secretName: kubernetes-dashboard-certs
+        emptyDir: {}
```
Don't merge this change. We require the dashboard bump and revert of this before merging into a release branch

joelsmith commented:
I'll add #57326 to this chery-pick PR as soon as it's squashed and merged.

joelsmith committed:
Fix nested volume mounts for read-only API data volumes

Since the runtime may try to create mount points within
the sandbox, it will fail if the mount point is within
a read-only API data volume, like a secret or configMap
volume.

Create any needed mount points during volume setup.

joelsmith commented:
We should wait until #60342 is done/merged and then add it here before considering this for merge.
/hold

joelsmith commented:
/kind bug
/sig storage
/sig node
/priority important-soon
/hold cancel
Canceling hold since #60342 is in the merge queue.

msau42 commented:
/lgtm

joelsmith commented:
/retest

joelsmith commented:
I didn't pay enough attention on the backport of the dashboard update. One of its changes prevented the dashboard from starting on 1.9:
```patch
-apiVersion: apps/v1beta2
+apiVersion: apps/v1
```
So I've reverted that change (and squashed the revert into the commit)

pull-kubernetes-verify should pass now that #60522 is merged
/retest

liggitt commented:
/lgtm

liggitt commented:
/status approved-for-milestone

smarterclayton commented:
/approve

mbohlool commented:
#58720 has a long release note. please update it with a shorter more user facing release note.

liggitt commented:
updated

joelsmith commented:
/test pull-kubernetes-e2e-kops-aws
